File: go/allergies/allergies.go
package allergies

const testVersion = 1

var allergyItems = []string{
	"eggs",
	"peanuts",
	"shellfish",
	"strawberries",
	"tomatoes",
	"chocolate",
	"pollen",
	"cats",
}

// Allergies lists the items that the allergy score indicates allergies to.
func Allergies(score uint) []string {
	allergies := []string{}

	for i, allergy := range allergyItems {
		if isScoreBit(score, i) {
			allergies = append(allergies, allergy)
		}
	}

	return allergies
}

// AllergicTo returns whether or not the provided allergy score indicates an allergy to the specified item.
func AllergicTo(score uint, item string) bool {
	for i, allergy := range allergyItems {
		if item == allergy {
			return isScoreBit(score, i)
		}
	}
	return false
}

func isScoreBit(score uint, b int) bool {
	return score&(1<<uint(b)) != 0
}

File: ruby/allergies/allergies.rb
# frozen_string_literal: true

class Allergies
  ALLERGENS = %w[
    eggs
    peanuts
    shellfish
    strawberries
    tomatoes
    chocolate
    pollen
    cats
  ].freeze

  def initialize(allergies)
    @allergies = allergies
  end

  def allergic_to?(allergen)
    raise unless bit = ALLERGENS.index(allergen)
    @allergies & (1 << bit) != 0
  end

  def list
    ALLERGENS.select { |a| allergic_to?(a) }
  end
end

module BookKeeping
  VERSION = 1
end
